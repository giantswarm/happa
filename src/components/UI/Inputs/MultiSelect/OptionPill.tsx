import { Box, Text } from 'grommet';
import PropTypes from 'prop-types';
import * as React from 'react';

interface IOptionPillProps extends React.ComponentPropsWithoutRef<typeof Box> {
  option: string;
  editable?: boolean;
}

const OptionPill: React.FC<IOptionPillProps> = ({
  option,
  editable,
  ...rest
}) => {
  return (
    <Box
      align='center'
      direction='row'
      gap='xsmall'
      pad={{ vertical: 'xsmall', horizontal: 'small' }}
      margin='xsmall'
      background='text-xweak'
      round='large'
      {...rest}
    >
      <Text size='small' color='text-strong'>
        {option}
      </Text>
      {editable && (
        <Box>
          <i className='fa fa-close' />
        </Box>
      )}
    </Box>
  );
};

OptionPill.propTypes = {
  option: PropTypes.string.isRequired,
  editable: PropTypes.bool,
};

OptionPill.defaultProps = {
  editable: false,
};

export default OptionPill;
