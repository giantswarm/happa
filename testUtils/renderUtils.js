import { render } from '@testing-library/react';
import App from 'App';
import { ConnectedRouter } from 'connected-react-router';
import { ThemeProvider } from 'emotion-theming';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { AppRoutes } from 'shared/constants/routes';
import configureStore from 'stores/configureStore';
import theme from 'styles/theme';

export const initialStorage = {
  user:
    '{"email":"developer@giantswarm.io","auth":{"scheme":"giantswarm","token":"a-valid-token"},"isAdmin":true}',
};

/**
 * This function will render the whole app with a mocked store in the route
 * provided.
 * @param {String} route.
 * @param {HTMLElement} container An HTMLElement.
 */
export function renderRouteWithStore(
  initialRoute = AppRoutes.Home,
  state = {},
  storage = initialStorage
) {
  localStorage.replaceWith(storage);

  const history = createMemoryHistory({
    initialEntries: [initialRoute],
    initialIndex: 0,
  });

  const store = configureStore(state, history);

  const app = render(<App {...{ store, theme, history }} />);

  return app;
}

/**
 * Render component in the Theme context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithTheme(component, props, options) {
  return render(getComponentWithTheme(component, props), options);
}

/**
 * Include component in the Theme context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 */
export function getComponentWithTheme(Component, props) {
  return (
    <ThemeProvider theme={theme}>
      <Component {...props} />
    </ThemeProvider>
  );
}

/**
 * Render component in the Store context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithStore(component, props, state, options) {
  return render(getComponentWithStore(component, props, state), options);
}

/**
 * Include component in the Store context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {Record<string, any>} storage Current LocalStorage value
 * @param {History} history Current Browser history
 */
export function getComponentWithStore(
  Component,
  props = {},
  state = {},
  storage = initialStorage,
  history = createMemoryHistory()
) {
  localStorage.replaceWith(storage);

  const store = configureStore(state, history);

  const app = (
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <ConnectedRouter history={history}>
          <Component {...props} />
        </ConnectedRouter>
      </ThemeProvider>
    </Provider>
  );

  return app;
}
