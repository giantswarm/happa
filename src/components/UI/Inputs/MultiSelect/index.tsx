import { Box, Button } from 'grommet';
import * as React from 'react';

import CheckBoxInput from '../CheckBoxInput';
import Select from '../Select';
import OptionPill from './OptionPill';

function getIDFromOption(option: string): string {
  return option.toLowerCase().replace(/\s/, '-');
}

interface IMultiSelectProps
  extends Omit<
    React.ComponentPropsWithoutRef<typeof Select>,
    'multiple' | 'value' | 'selected'
  > {
  /**
   * The values that can be chosen from the dropdown.
   */
  options: string[];
  /**
   * The maximum number of selected values that can
   * be seen in the top view.
   */
  maxVisibleValues?: number;
  /**
   * The function that gets called when a value
   * is deleted.
   */
  onRemoveValue?: (option: string) => void;
  /**
   * The options that are currently selected.
   */
  selected?: string[];
  /**
   * Whether the input should be disabled or not.
   */
  disabled?: boolean;
}

const MultiSelect = React.forwardRef<HTMLButtonElement, IMultiSelectProps>(
  (
    {
      placeholder,
      onRemoveValue,
      selected,
      options,
      maxVisibleValues,
      disabled,
      ...props
    },
    ref
  ) => {
    const visibleSelections = selected?.slice(0, maxVisibleValues) ?? [];
    const extraSelectionsLength = (selected?.length ?? 0) - maxVisibleValues!;

    const selectedOptionsIdx = React.useMemo(() => {
      return options.reduce<number[]>((acc, curr, idx) => {
        if (selected?.includes(curr)) {
          acc.push(idx);
        }

        return acc;
      }, []);
    }, [selected, options]);

    return (
      <Select
        multiple={true}
        closeOnChange={false}
        placeholder={placeholder}
        selected={selectedOptionsIdx}
        options={options}
        disabled={disabled}
        value={
          visibleSelections.length > 0 ? (
            <Box wrap={true} direction='row'>
              {visibleSelections.map((option) => (
                <Button
                  key={`button_${option}`}
                  href='#'
                  disabled={disabled}
                  onClick={(e) => {
                    e.preventDefault();
                    e.stopPropagation();

                    if (disabled) return;

                    onRemoveValue?.(option);
                  }}
                  onFocus={(e) => e.stopPropagation()}
                >
                  <OptionPill option={option} editable={!disabled} />
                </Button>
              ))}

              {extraSelectionsLength > 0 && (
                <OptionPill
                  option={`+${extraSelectionsLength} more`}
                  editable={false}
                />
              )}
            </Box>
          ) : undefined
        }
        {...props}
        ref={ref}
      >
        {(option, _idx, _options, state) => {
          return (
            <CheckBoxInput
              key={option}
              id={getIDFromOption(option)}
              name={getIDFromOption(option)}
              tabIndex={-1}
              checked={state.selected}
              disabled={state.disabled}
              label={option}
              margin='none'
              contentProps={{
                pad: 'small',
              }}
            />
          );
        }}
      </Select>
    );
  }
);

MultiSelect.defaultProps = {
  maxVisibleValues: 4,
};

export default MultiSelect;
