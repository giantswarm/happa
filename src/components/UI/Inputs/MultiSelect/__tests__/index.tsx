import { fireEvent, screen } from '@testing-library/react';
import { renderWithTheme } from 'test/renderUtils';

import MultiSelect from '..';

describe('MultiSelect', () => {
  it('renders a simple input', () => {
    const { container } = renderWithTheme(MultiSelect, {
      id: 'pet',
      options: [
        'A dog',
        'A cat',
        'A frog',
        'A shark',
        'A ghost',
        'A snake',
        'A camel',
      ],
      placeholder: 'Just select something...',
      label: 'Pets I want to get',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('displays all the selected values, and a placeholder if the number of selected values is above the given limit', () => {
    renderWithTheme(MultiSelect, {
      id: 'pet',
      options: [
        'A dog',
        'A cat',
        'A frog',
        'A shark',
        'A ghost',
        'A snake',
        'A camel',
      ],
      placeholder: 'Just select something...',
      label: 'Pets I want to get',
      selected: ['A dog', 'A cat', 'A frog', 'A shark'],
      maxVisibleValues: 2,
    });

    expect(screen.getByText('A dog')).toBeInTheDocument();
    expect(screen.getByText('A cat')).toBeInTheDocument();
    expect(screen.getByText('+2 more')).toBeInTheDocument();
  });

  it('can remove selected values', () => {
    const onRemoveValueMockFn = jest.fn();

    renderWithTheme(MultiSelect, {
      id: 'pet',
      options: [
        'A dog',
        'A cat',
        'A frog',
        'A shark',
        'A ghost',
        'A snake',
        'A camel',
      ],
      placeholder: 'Just select something...',
      label: 'Pets I want to get',
      selected: ['A dog', 'A cat', 'A frog', 'A shark'],
      maxVisibleValues: 2,
      onRemoveValue: onRemoveValueMockFn,
    });

    const value = screen.getByText('A cat');
    expect(value).toBeInTheDocument();
    fireEvent.click(value);

    expect(onRemoveValueMockFn).toHaveBeenCalledWith('A cat');
  });
});
