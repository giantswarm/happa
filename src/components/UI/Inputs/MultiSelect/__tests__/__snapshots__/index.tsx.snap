// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect renders a simple input 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 hSiATR"
>
  <div
    class="ThemeProvider__AppWrapper-sc-1qc9t9l-0 bLmXgB"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 erPdoW FormField__FormFieldBox-m9hood-0 gWgpjB"
    >
      <label
        class="StyledText-sc-1sadyjn-0 weZKr"
        for="pet"
      >
        Pets I want to get
      </label>
      <div
        class="StyledBox-sc-13pk1d4-0 gcjayZ FormField__FormFieldContentBox-m9hood-1"
      >
        <button
          aria-label="Open Drop"
          class="StyledButton-sc-323bzc-0 ffRhrh Select__StyledSelectDropButton-sc-17idtfo-1 cmdTsS"
          id="pet"
          type="button"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 ftwAxs"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 dtQiSn"
            >
              <div
                class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 cnyLED"
              >
                <input
                  autocomplete="off"
                  class="StyledTextInput-sc-1x30a0s-0 hPlSXi Select__SelectTextInput-sc-17idtfo-0 kMNVoA"
                  id="pet__input"
                  placeholder="Just select something..."
                  readonly=""
                  tabindex="-1"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="StyledBox-sc-13pk1d4-0 cdaeaF"
              style="min-width: auto;"
            >
              <svg
                aria-label="FormDown"
                class="StyledIcon-ofa7kd-0 gJBwzH"
                viewBox="0 0 24 24"
              >
                <polyline
                  fill="none"
                  points="18 9 12 15 6 9"
                  stroke="#000"
                  stroke-width="2"
                />
              </svg>
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
